wolfcrypt/src/port/st/STM32MP13.md

```diff
@@ -0,0 +1,212 @@
+`printf()`, then you need to set PD6 and PD8 as the UART 4 RX/TX pins. You can
+then enable UART4 and set it to "Asynchronous" mode.
+
+In the code 0 section of `main.c` add:
+
+```c
+#ifdef __GNUC__
+int __io_putchar(int ch)
+#else
+int fputc(int ch, FILE *f)
+#endif
+{
+    HAL_UART_Transmit(&huart4, (uint8_t *)&ch, 1, 0xFFFF);
+
+    return ch;
+}
+#ifdef __GNUC__
+int _write(int file,char *ptr, int len)
+{
+    int DataIdx;
+    for (DataIdx= 0; DataIdx< len; DataIdx++) {
+        __io_putchar(*ptr++);
+    }
+    return len;
+}
+#endif
+```
+
+UART4 will now be used for `printf()`.
+
+
+
+### wolfSSL in your project
+
+There are a few things you need to do to get wolfSSL to run in your project. The
+first is setting compile option, these additional ones are needed. The first
+allows ARM ASM optimizations to compile, the second stops alignment issues from
+crashing the board:
+
+```
+-fomit-frame-pointer
+-mno-unaligned-access
+```
+
+The first of these should also be a flag for the assembler as well.
+
+Then the code needs to be set to use the DDR RAM instead of SRAM. To do this,
+edit `STM32MP135FAFX_RAM.ld` and change:
+
+```c
+REGION_ALIAS("RAM", SYSRAM_BASE);
+```
+
+To this:
+
+```c
+REGION_ALIAS("RAM", DDR_BASE);
+```
+
+In the Run Configuration menu, make sure that the debugger's startup has the
+"monitor reset" command removed. Otherwise the DDR initialization will be reset.
+
+In the `main.c` make sure that `SystemClock_Config();` is not executed. The DDR
+Init code will do this, and changing it will likely crash the board. It can
+be done like this:
+
+```c
+  /* USER CODE BEGIN Init */
+#if 0
+  /* USER CODE END Init */
+
+  /* Configure the system clock */
+  SystemClock_Config();
+
+  /* USER CODE BEGIN SysInit */
+#endif
+  /* USER CODE END SysInit */
+```
+
+### Benchmark
+
+To use the wolfCrypt benchmark, enable and activate the RTC in the Device
+Configuration Tool and then add this to your `main.c`:
+
+```c
+double current_time(void)
+{
+    RTC_TimeTypeDef time;
+    RTC_DateTypeDef date;
+    uint32_t subsec = 0;
+
+    /* must get time and date here due to STM32 HW bug */
+    HAL_RTC_GetTime(&hrtc, &time, RTC_FORMAT_BIN);
+    HAL_RTC_GetDate(&hrtc, &date, RTC_FORMAT_BIN);
+    /* Not all STM32 RTCs have subseconds in the struct */
+#ifdef RTC_ALARMSUBSECONDMASK_ALL
+    subsec = (255 - time.SubSeconds) * 1000 / 255;
+#endif
+
+    (void) date;
+
+    /* return seconds.milliseconds */
+    return ((double) time.Hours * 24) + ((double) time.Minutes * 60)
+            + (double) time.Seconds + ((double) subsec / 1000);
+}
+```
+
+### Compiling wolfSSL
+
+In your `user_settings.h` you should include:
+
+```c
+#define WOLFSSL_STM32MP13
+#define WOLFSSL_STM32_CUBEMX
+#define WOLFSSL_USER_CURRTIME
+```
+
+If you want ECDSA acceleration, you should also add:
+
+```c
+#define WOLFSSL_STM32_PKA
+#define WOLFSSL_STM32_PKA_V2
+```
+
+### Running
+
+Once you have compiled everything, to run your code you will first need to run
```

Comment:
Consider manually including the DDR init code required in the example. This README.md does not include steps for adding the calls to wolfCrypt_test / bench. That usually come with `IDE/STM32Cube/wolfssl_example.c`

Reply:
We would only be able to integrate the DDR Init code if we were running from the SD card (I don't know how to do it yet, but it must be possible that way), that requires signed binaries and is not a good place to do development / debugging.

Good point on the calls. I'll see what I can do there.